### DISCUSSION TITLE
Help with creating new header / modifying rp6502-vscode example(s) (for novices)

### DISCUSSION THREAD

**Discussion by Hyenadae** [picocomputer/community#65] [Q&A] [answered]
Hey again RP6502 community,

Something that was bothering me yesterday was that there wasn't a "complete doofus with goldfish memory" sort of library/header for us to use when doing the initialization of various graphical modes. Given the examples from both github and video, it'd be nice if you could 'just' call a function for any given init work, and what would be having 5-10 lines of specific init code in a certain order and values would be simplified to one or two calls.

I tried making something really basic, base off of the provided examples but VSCode / Cmake isn't being helpful in seeing the new header file, that's right next to the base (rp6502.h) one in the folder. 

Here's the code so far:

**hello.c modified**
```

#include <rp6502.h>
#include <stdio.h>
#include <rp6502_easy_gfx.h>

void main()
{

    puts("Hello, world!");
    rp6502_gfx_mode3_init(); 

}

```

**New   rp6502_easy_gfx.h  file in  rp6502-vscode/include**
```
#ifndef _RP6502_EASY_GFX_H_
#define _RP6502_EASY_GFX_H_
#include <rp6502.h>

#define rp6502_gfx_mode3_clear(void)    \ /* Assuming we want to clear most of XRAM, for hi-color, low-res Mode3 */
    unsigned i;                         \
    RIA.addr0 = 0;                      \
    RIA.step0 = 1;                      \
    for (i = 0; i < 61440u / 8; i++)    \
    {                                   \
        RIA.rw0 = 0;                    \
        RIA.rw0 = 0;                    \
        RIA.rw0 = 0;                    \
        RIA.rw0 = 0;                    \
        RIA.rw0 = 0;                    \
        RIA.rw0 = 0;                    \
        RIA.rw0 = 0;                    \
        RIA.rw0 = 0;                    \
    }                                   


#define rp6502_gfx_mode3_init(void) \
    rp6502_gfx_mode3_clear();                                   \   /* Call Clear XRAM function*/
    xreg_vga_canvas(2);                                         \   /* Set Canvas mode to 2 (320x180)*/      
    xram0_struct_set(0xFF00, vga_mode3_config_t, x_wrap, true); \
    xram0_struct_set(0xFF00, vga_mode3_config_t, y_wrap, true); \
    xram0_struct_set(0xFF00, vga_mode3_config_t, x_pos_px, 0); \
    xram0_struct_set(0xFF00, vga_mode3_config_t, y_pos_px, 0); \
    xram0_struct_set(0xFF00, vga_mode3_config_t, width_px, 240); \
    xram0_struct_set(0xFF00, vga_mode3_config_t, height_px, 128); \
    xram0_struct_set(0xFF00, vga_mode3_config_t, xram_data_ptr, 0x0000); \
    xram0_struct_set(0xFF00, vga_mode3_config_t, xram_palette_ptr, 0xFFFF); \
    xreg(1, 0, 1, 3, 2, 0xFF00); 


#endif /* _RP6502_EASY_GFX_H_ */
```

Long story short:
[build] [ 50%] Building C object CMakeFiles/hello.dir/src/main.c.obj
[build] /home/hy/src/pico6502/rp6502-vscode/src/main.c:3: Error: Include file 'rp6502_easy_gfx.h' not found
[build] /home/hy/src/pico6502/rp6502-vscode/src/main.c:10: Error: Call to undeclared function 'rp6502_gfx_mode3_init'
[build] make[2]: *** [CMakeFiles/hello.dir/build.make:63: CMakeFiles/hello.dir/src/main.c.obj] Error 1
[build] make[1]: *** [CMakeFiles/Makefile2:96: CMakeFiles/hello.dir/all] Error 2

It'd be nice to understand how header files are properly made, and how not to do XYZ as well, but getting to debugging the header file is hard when said file can't be seen. I've never edited CMAKE configs either, so any advice (and why it doesn't look through *all of include folder*) would be nice

**Comment by Hyenadae** [comment]
I've been referencing Tetricks https://github.com/orgs/picocomputer/discussions/54 
 and they seem to shove the headers into the src folder with quotations (which I know is a way to force it... but still).

Additionally, they seem to have all the init code not 'abstracted' away to a header, and have it in its normal form in the C file. 
It does have nice functions though as a reference, but it seems a bit 'heavy' with all the shape drawing functions, so I might want to strip out the simple stuff.

Compilation for the 'header' in the src directory and properly found by the compiler/hello.c is giving plenty of errors as I don't know how to properly set it up, lol.

**Comment by vruumllc** [comment]
I already did all this work. Check out the bitmap_graphics_library

**Reply by Hyenadae** [reply]
So using the example of erase_canvas which doesn't take in any information to work as a function, or give out any info,

All you need to do to make a globally accessible function is to:
1. first create it in another C file (actual code goes here), 
2. then make a header that keeps defining the function as void func(void) (as a shim/glue) and 
3. with the 'main' C file including the header (not the actual file with the original source of the executing code)
4. do the usual func(); call when you need it, and that function will still have access to all variables that are active if it needs any, as long as you don't re-init them within the function(?)
5. Edit Cmake lists to include both C source files and make sure the header is besides them, then compile

**Comment by netzerohero** [comment, marked as answer]
In response to your original post at the top:
It appears C isn't your native language.   /grin.
And IDEs like vscode with Cmake tend to hide users with all their "magic performed under-the-hood". 
(Hence your comment about difficulty-to-debug header-file since the IDE isn't following your as written C-syntax.)  /grin.

1. For include files in main.c there is a difference between:
`#include <rp6502_easy_gfx.h>`
and
`#include "./rp6502_easy_gfx.h" `
Or try: 
`#include "rp6502_easy_gfx.h" `

2. For your new rp6502_easy_gfx.h file:
There are a couple new-to-C isssues...
The two new functions you've attempted to define are syntactically problematic.  As currently written, the compiler pre-pass is interpreting those two long "#defines" as pre-compiler macros.  I don't believe you intended that. Additionally, your two new functions have not been declared. 

Rather, try:
   a. Rename the rp6502_easy_gfx.h to an rp6502_easy_gfx.c file; 
   b. Remove the top two lines and your bottom #endif.
```
   #ifndef _RP6502_EASY_GFX_H_
   #define _RP6502_EASY_GFX_H_
```
   `#endif  `

 ...Those are multi-pass include protections unnecessary on a *.c source file.

c.  Change the two following from:
```
   #define rp6502_gfx_mode3_clear(void) \
   #define rp6502_gfx_mode3_init(void) \
```
   to:
```
   void rp6502_gfx_mode3_clear(void) 
   void rp6502_gfx_mode3_init(void)
```

d. Remove all those single backslashes (if they are indeed single backslashes and not editor artifacts / issues in your post) at the end of every line of your two function definitions.  If you want comments in c, be sure to either use two consecutive backslashes as:
`\\ my comment at end of my code line`
or enclose comment as you have done:
`/* my comment between these two special characters */ `

d. Remember to declare your two new c-functions near the top of your file, before the 1st function define:
```
void rp6502_gfx_mode3_clear(void); 
void rp6502_gfx_mode3_init(void);
```

Moreover, expanding on what vruumllc said concisely above... 
Suggest going back and reviewing his well-formed c/h source code packaging patterns.  Once your c-file (as I've suggested you modify) plays-nice, you can follow vruumllc's pattern and break-up and re-package your code into a c/h pair as he has done.

The above should allow you to better accomplish what you desired.  (I didn't try this on my dev system; I'm just reading your code posted.) 

Hope this helps.

**Comment by Hyenadae** [comment]
Thanks LOL

Last time I coded in a C variant was a mixture of C and OpenCL two years ago, and the definitions and methods of sharing data can be quite different because of how detached and stupid the GPU is. But the bitmap library and its lack of "fully stuffed" headers made me realize how things work a bit differently and how to actually make functions of working code that you can recall, as I had expected them to work based on my cluelessness haha
